### PR TITLE
Speed up reading and merging many files

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

Reading and merging spools with many (hundreds of) files was much slower than the underlying file IO: profiling a 300-file merge showed only ~30% of the time going to actual HDF5 reads, the rest being fixed per-file python overhead (~12 ms/file). Merging also held every source patch plus the concatenated copy in memory at once. This PR removes the biggest overheads in the load path and streams merges:

- `DirectorySpool._load_patch` reads directly through the file's `FiberIO`, skipping `dc.read` format detection and the spool indexing machinery applied to its throwaway output. Range selection is only applied when the patch is actually trimmed (tracked by the instruction dataframe's `_modified` flag). The fast path requires a concrete format/version and falls back to `dc.read` otherwise (including for multi-patch files without selection, so each row still maps to its own patch).
- `MemorySpool` is lazy: the managing dataframes are only built when an operation needs them, and `len`, integer access, and iteration are served directly from an immutable patch tuple for spools created from patches. This makes the spool created (and immediately unwrapped) by every `FiberIO.read` call nearly free. Equality is normalized so it doesn't depend on whether the lazy dataframes were built, derived spools don't retain their parents, and `new_from_df` no longer builds dummy dataframes it discards.
- **Merges stream into a pre-allocated array** sized from the instruction dataframe instead of loading all sources and concatenating: peak memory for a merge drops ~45% (to output + one file) and a full-data copy is removed. Mixed dtypes upcast as `np.concatenate` would; inexact estimates grow or trim the buffer; inestimable cases (unknown steps, no varying indexed dim) use the previous materialized merge. Both paths produce identical results, and gap/continuity checks are unchanged (they happen on the dataframes at chunk time, before any data is read).
- `_get_patches_from_index` looks up instruction rows via a cached `groupby` mapping instead of scanning the whole instruction dataframe per patch, removing O(sources × outputs) scaling when iterating chunked spools.
- Patches trimmed to nothing (see #583) raise a dedicated `MissingPatchError` (an `IndexError` subclass for backwards compatibility) which iteration catches by type; the previous "out of bounds" message matching is gone and unexpected `IndexError`s always propagate. `FileSpool` had the same latent crash and gets the same treatment.
- `get_quantity_str` (called by the pydantic serializers of attrs/coord models on every dump) caches string validation and unit stringification. Caches are only keyed on types with exact equality (`str` and `pint.Unit`); a `Quantity` cannot key a cache because equal quantities with different units (`1 m == 100 cm`) would collide. A test guards this.
- New CodSpeed benchmarks cover indexing, merging, and chunked iteration over a many-file directory, plus in-memory spool creation/access and many-patch merges (~1 s total wall time).

Benchmarks (DASDAE format, warm cache, local disk):

| Scenario | Before | After |
|---|---|---|
| 300 files (1.4 GB), `chunk(time=None)` merge | ~4.6 s | ~1.9 s |
| — peak memory for the above | ~3.0 GB | ~1.6 GB |
| 1000 small files, merge all | ~12 s | ~3.4 s |
| 1000 small files, chunk into 250 patches | ~15 s | ~4.7 s |

For reference, raw pytables reads + `np.concatenate` of the same 300-file dataset take ~1.5 s. A likely follow-up (separate PR) is optional executor-parallel patch loading for cold/network storage.

## Checklist

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [x] documented the new feature with docstrings and/or appropriate doc page.
- [x] included tests. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [x] added the "ready_for_review" tag once the PR is ready to be reviewed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Faster patch loading for directory-based workflows with improved handling of multi-patch files.
  * Enhanced lazy in-memory spooling for more efficient access patterns.
* **Bug Fixes**
  * Missing/trimmed entries now raise `MissingPatchError` (still catchable as `IndexError`), improving patch disambiguation.
  * Chunk/merge behavior is more consistent: dimension order and coordinate metadata are preserved; streaming merge matches materialized output.
  * Quantity-to-string formatting avoids over-normalizing equal quantities.
* **Tests**
  * Added coverage for multi-patch directory behavior, lazy spool semantics, streaming merge edge cases, and unit formatting.

### CODE REVIEW

File: dascore/clients/dirspool.py [hunk 1]
```diff
@@ -126,5 +126,46 @@ def _load_patch(self, kwargs) -> Self:
...
+            select = {}
+        spool = fiber_io.read(kwargs["path"], **select)
+        if not isinstance(spool, MemorySpool):
+            return None
+        patches = spool._unprocessed_patches()
+        # Without selection, a multi-patch file is ambiguous: the row refers
+        # to one specific patch. Let the generic path pick the right one.
+        if len(patches) > 1 and not select:
+            return None
+        return patches
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/clients/dirspool.py:175]: Good catch — the fast path now requires a concrete format and version and otherwise defers to dc.read, which detects the actual version from the file. Fixed in e0c8916d, with a regression test.

File: dascore/units.py [hunk 2]
```diff
@@ -220,22 +231,34 @@ def get_quantity_str(quant_value: str | Quantity | None) -> str | None:
...
 
 
+@cache
+def _validate_quantity_str(quant_str: str) -> None:
+    """Raise a UnitError if the string doesn't specify a valid quantity."""
+    try:
+        get_quantity(quant_str)
+    except UndefinedUnitError:
+        msg = f"DASCore failed to parse the following unit/quantity: {quant_str}"
+        raise UnitError(msg)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/units.py]: Fixed in e0c8916d (raise ... from e).

File: dascore/core/spool.py [hunk 3]
```diff
@@ -665,13 +667,101 @@ def get_contents(self) -> pd.DataFrame:
...
+        if (
+            patches is not None
+            and not self._select_kwargs
+            and isinstance(item, int | np.integer)
+        ):
+            try:
+                return patches[item]
+            except IndexError:
+                msg = f"index of [{item}] is out of bounds for spool."
+                raise IndexError(msg)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/spool.py]: Fixed in e0c8916d (raise ... from None).

File: dascore/core/spool.py [hunk 4]
```diff
@@ -53,6 +62,62 @@
 T = TypeVar("T")
 
 
+def _dim_slice(ndim: int, axis: int, start: int, stop: int) -> tuple:
```
Comment (d-chambers) [inline dascore/core/spool.py]: There is already a utility for this, `broadcast_for_index`, just use that.

File: dascore/core/spool.py [hunk 5]
```diff
@@ -53,6 +62,62 @@
...
+    """Get an nd slice tuple selecting start:stop along axis."""
+    out = [slice(None)] * ndim
+    out[axis] = slice(start, stop)
+    return tuple(out)
+
+
+def _patches_from_read(spool: BaseSpool) -> list[PatchType]:
+    """Get the raw patch list from a spool returned by dc.read."""
+    if isinstance(spool, MemorySpool):
```
Comment (d-chambers) [inline dascore/core/spool.py]: Do we really need this helper? Shouldn't the list transformation do this already then we can just remove this code?

File: dascore/core/spool.py [hunk 6]
```diff
@@ -372,6 +441,15 @@ def _get_source_df(self):
     def _get_instruction_df(self):
         """Function to get the current df."""
 
+    def _get_instruction_indices(self):
```
Comment (d-chambers) [inline dascore/core/spool.py]: Can we just inline this?

File: dascore/core/spool.py [hunk 7]
```diff
@@ -663,15 +825,106 @@ def get_contents(self) -> pd.DataFrame:
...
         if data is not None:
-            dfs = self._get_dummy_dataframes(patches_to_df(data))
-            self._df, self._source_df, self._instruction_df = dfs
+            if isinstance(data, dc.Patch):
+                self._patches = (data,)
+            elif isinstance(data, Sequence) and all(
+                isinstance(x, dc.Patch) for x in data
+            ):
+                # Copy to an immutable tuple so neither mutating the input
+                # sequence nor the stored one can change the spool contents.
```
Comment (d-chambers) [inline dascore/core/spool.py]: remove this comment

File: dascore/core/spool.py [hunk 8]
```diff
@@ -692,6 +945,26 @@ def _load_patch(self, kwargs) -> Self:
         """Load the patch into memory."""
         return kwargs["patch"]
 
+    def _unprocessed_patches(self) -> list[PatchType]:
```
Comment (d-chambers) [inline dascore/core/spool.py]: This has a bed smell to it. SHouldn't Spool.__iter__ already handle this such that `list(spool)` would do the same thing as this?

File: dascore/utils/patch.py [hunk 9]
```diff
@@ -445,14 +454,40 @@ def _get_new_coord(df, merge_dim, coords, drop_conflicting=False):
...
-    new_coord = _get_new_coord(df, merge_dim, coords, drop_conf_coords)
+    new_coord = _get_merged_coord(df, merge_dim, coords, drop_conf_coords)
     coord = new_coord.coord_map[merge_dim] if merge_dim in dims else None
     new_attrs = combine_patch_attrs(attrs, merge_dim, coord=coord, **merge_kwargs)
     patch = dc.Patch(data=new_data, coords=new_coord, attrs=new_attrs, dims=dims)
     new_dict = {"patch": patch}
     return [new_dict]
 
 
+def _patch_dim_summary(patch: PatchType) -> dict:
```
Comment (d-chambers) [inline dascore/utils/patch.py]: This should probably be a method on the `CoordManager`

File: dascore/core/spool.py (hunk 4 above)
Comment (d-chambers) [reply dascore/core/spool.py]: Done in a43da4ca — replaced with broadcast_for_index.

File: dascore/core/spool.py (hunk 5 above)
Comment (d-chambers) [reply dascore/core/spool.py]: Right — with the lazy iteration fast path, list(spool) already returns the raw patches, so both this helper and _unprocessed_patches are removed in a43da4ca.

File: dascore/core/spool.py (hunk 6 above)
Comment (d-chambers) [reply dascore/core/spool.py]: Inlined into _get_patches_from_index with a manual cache lookup in a43da4ca.

File: dascore/core/spool.py (hunk 7 above)
Comment (d-chambers) [reply dascore/core/spool.py]: Removed in a43da4ca.

File: dascore/core/spool.py (hunk 8 above)
Comment (d-chambers) [reply dascore/core/spool.py]: Agreed — the lazy __iter__ fast path makes list(spool) equivalent, so this method is removed and all callers use list(spool) (a43da4ca).

File: dascore/utils/patch.py (hunk 9 above)
Comment (d-chambers) [reply dascore/utils/patch.py]: Moved to CoordManager._get_dim_summary in a43da4ca.
